forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt

```diff
@@ -0,0 +1,191 @@
+    }
+
+    /**
+     * Helper method to execute signing, show biometric prompt.
+     *
+     * @param userKey User Information
+     * @param listener The Listener to listen for the result
+     * @param authInterface Interface to find the Authentication Type
+     */
+    protected open fun authenticate(userKey: UserKey,
+                                    listener: FRListener<Void>,
+                                    authInterface: Authenticator = getAuthenticator(userKey)) {
+
+        if(authInterface.isSupported().not()) {
+            handleException(Unsupported(), listener)
+            return
+        }
+        try {
+            authInterface.authenticate(timeout ?: 60) { result ->
+                if (result is Success) {
+                    val jws = authInterface.sign(userKey, challenge)
+                    setJws(jws)
+                    Listener.onSuccess(listener, null)
+                } else {
+                    // All the biometric exception is handled here , it could be Abort or timeout
+                    handleException(result, listener = listener)
+                }
+            }
+        }
+        catch (e: Exception) {
+            // This Exception happens only when there is Signing or keypair failed.
+            handleException(Unsupported(errorMessage = e.message), listener = listener)
+        }
+    }
+
+    /**
+     * Create the interface for the Authentication type(Biometric, Biometric_Fallback, none)
+     * @param context application or activity context
+     * @param userKeyService service to sort and fetch the keys stored in the device
+     * @param listener The Listener to listen for the result
+     */
+    protected open fun execute(context: Context,
+                               userKeyService: UserKeyService = UserDeviceKeyService(context),
+                               listener: FRListener<Void>) {
+
+        when(val status = userKeyService.getKeyStatus(userId)) {
+            is NoKeysFound -> handleException(UnRegister(), listener)
+            is SingleKeyFound -> authenticate(status.key , listener = listener)
+            else -> { showKeysFragment(InitProvider.getCurrentActivityAsFragmentActivity(), userKeyService, listener =  listener) }
+        }
+    }
+
+    /**
+     * Display fragment to select a user key from the list
+     * @param activity activity to be used to display the Fragment
+     * @param userKeyService service to sort and fetch the keys stored in the device
+     * @param fragment to display the users list
+     * @param listener The Listener to listen for the result
+     */
+    protected open fun showKeysFragment(activity: FragmentActivity,
+                                        userKeyService: UserKeyService = UserDeviceKeyService(activity),
+                                        fragment: DialogFragment =  DeviceBindFragment(userKeyService),
+                                        listener: FRListener<Void>) {
+
+        userKeyService.callback = {
```

Comment:
Do we need the callback to associate with the UserKeyService?

Reply:
Fragment gets the userkeys from UserKeyService and set a selected key to UserKeyService and notify the Callback.  This is Similar to LiveData approach . we can also achieve this in multiple ways , we can also Directly fetch selected item from Fragment which may look cleaner

Reply:
Understand the idea, but reset the callback may be a concern, if the caller providers the callback for the userKeyService, the callback may be overwritten by this method.

Reply:
this is taken care